File: examples/py/08_predict_mc_learning.py
"""Example 08: Statistical Prediction — Learned / Monte Carlo

Demonstrates predict() when no analytical distribution transfer is known.
The framework automatically learns the output distribution by sampling through
the node functions (Monte Carlo forward pass).

Pipeline
--------
    x ~ Normal(μ=0, σ=1)
        │
    [NonLinear]   y = sin(x) · exp(−x²/4)            (no closed form)
        │
    [Rectify]     z = max(0, y)  (half-wave rectifier) (no closed form)
        │
    [Scale]       w = 3z + 1                            (linear — attach transfer)
        │
    [Square]      out = w²                              (no closed form)

Key concepts shown:
  - Full MC pass with no transfers  → all Empirical outputs
  - Mixed-mode: analytical Scale, MC for the rest
  - Effect of n_samples on accuracy
  - Accessing output distributions: mean, std, percentiles, raw samples
  - at_node early stop to inspect intermediate variables cheaply
"""

import sys
sys.path.insert(0, '/home/runner/work/graph-sp/graph-sp/examples/py')

import math
from benchmark_utils import print_header, print_section, print_dist_table, Benchmark
import dagex


# ── Node functions ────────────────────────────────────────────────────────────

def non_linear(inputs):
    """y = sin(x) * exp(-x²/4)  — a Mexican-hat-like pulse."""
    import math
    x = inputs.get("x", 0)
    return {"y": math.sin(x) * math.exp(-(x ** 2) / 4.0)}


def rectify(inputs):
    """z = max(0, y)  — half-wave rectification (keeps positive part only)."""
    y = inputs.get("y", 0)
    return {"z": max(0.0, y)}


def scale(inputs):
    """w = 3z + 1  — linear scaling."""
    z = inputs.get("z", 0)
    return {"w": z * 3.0 + 1.0}


def square(inputs):
    """out = w²."""
    w = inputs.get("w", 0)
    return {"out": w ** 2}


def main():
    print_header("Example 08: Statistical Prediction — Monte Carlo Learning")

    print("📖 Story:")
    print("   A non-linear signal processing pipeline: input noise passes through")
    print("   a Mexican-hat nonlinearity, half-wave rectification, a linear scale,")
    print("   and squaring.  Only the linear Scale node has a known analytical")
    print("   distribution transfer.  All others are learned via Monte Carlo.\n")
    print("   This shows how predict() works transparently even without any domain")
    print("   knowledge — just supply the prior on the input and let it sample.\n")

    # ── Build graph ───────────────────────────────────────────────────────────
    print_section("Building the Graph")

    graph = dagex.Graph()
    graph.add(non_linear, label="NonLinear", inputs=[("x",  "x")], outputs=[("y", "y")])
    graph.add(rectify,    label="Rectify",   inputs=[("y",  "y")], outputs=[("z", "z")])
    graph.add(scale,      label="Scale",     inputs=[("z",  "z")], outputs=[("w", "w")])
    graph.add(square,     label="Square",    inputs=[("w",  "w")], outputs=[("out", "out")])

    # Only Scale has a known distribution transfer
    def scale_dist(dists):
        z = dists["z"]
        return {"w": dagex.normal(z.mean * 3.0 + 1.0, z.std * 3.0)}

    graph.set_dist_transfer("Scale", scale_dist)

    dag = graph.build()
    print(dag.to_mermaid())

    x_prior = {"x": dagex.normal(0.0, 1.0)}

    # ── Full MC pass — no transfers at all first ──────────────────────────────
    print_section("Full Monte Carlo (no transfers, n_samples=500)")

    graph_plain = dagex.Graph()
    graph_plain.add(non_linear, label="NonLinear", inputs=[("x", "x")], outputs=[("y", "y")])
    graph_plain.add(rectify,    label="Rectify",   inputs=[("y", "y")], outputs=[("z", "z")])
    graph_plain.add(scale,      label="Scale",     inputs=[("z", "z")], outputs=[("w", "w")])
    graph_plain.add(square,     label="Square",    inputs=[("w", "w")], outputs=[("out", "out")])
    dag_plain = graph_plain.build()

    with Benchmark("MC n=500") as bm:
        stat_plain = dag_plain.predict(x_prior, n_samples=500)
    print(f"  Time: {bm.result.duration_ms:.1f} ms")
    print(f"  out:  {stat_plain['out'].summary()}")

    # ── Mixed-mode (Scale analytical, rest MC) ────────────────────────────────
    print_section("Mixed-mode (Scale analytical + MC elsewhere)")

    for n in [200, 1000, 5000]:
        with Benchmark(f"n={n}") as bm:
            stat = dag.predict(x_prior, n_samples=n)
        d_out = stat["out"]
        d_w   = stat["w"]
        print(f"  n={n:5d}  out: mean={d_out.mean:.4f}  std={d_out.std:.4f}"
              f"  [p5={d_out.p5:.3f}, p95={d_out.p95:.3f}]"
              f"  ({bm.result.duration_ms:.0f} ms)")
        if n == 5000:
            print(f"         w (analytical Scale input):  {d_w}")

    print()
    print("  Note: w comes from the analytical Scale transfer (Normal).")
    print("  Upstream/downstream nodes are Empirical (MC samples).\n")

    # ── Inspecting individual distributions ───────────────────────────────────
    print_section("Inspecting intermediate distributions (n_samples=3000)")

    stat = dag.predict(x_prior, n_samples=3000)

    rows = []
    for var in ["x", "y", "z", "w", "out"]:
        d = stat[var]
        type_str = repr(d).split("(")[0]
        rows.append((var, d, "", type_str))
    print_dist_table(rows, show_type=True)

    print()
    print("  Empirical types can be histogrammed:")
    z_samples = stat["z"].samples
    if z_samples:
        positive = sum(1 for s in z_samples if s > 0) / len(z_samples)
        print(f"  z > 0: {positive:.1%}  (expected ≈50% before rectification,")
        print(f"         actual >{positive:.0%} because rectify clips z to ≥0)")

    # ── Early stop: only compute up to Rectify ───────────────────────────────
    print_section("at_node='Rectify' — early stop before Scale and Square")

    stat_early = dag.predict(x_prior, n_samples=2000, at_node="Rectify")
    print("  Keys:", stat_early.keys())
    print("  'w' present:   ", stat_early.get("w")   is not None)
    print("  'out' present: ", stat_early.get("out") is not None)
    print("  z (Rectify output):", stat_early["z"].summary())

    # ── Compare distributions across sample counts ────────────────────────────
    print_section("Effect of n_samples on output mean estimate")

    print("  n_samples   out.mean   out.std")
    print("  " + "-" * 35)
    for n in [50, 200, 1000, 5000]:
        s = dag.predict(x_prior, n_samples=n)
        d = s["out"]
        print(f"  {n:8d}   {d.mean:8.4f}   {d.std:8.4f}")

    print()
    print("  As n_samples increases, estimates converge.")

    # ── Joint distribution analysis ───────────────────────────────────────────
    print_section("Joint Distribution Analysis  (predict_particles)")

    print("  Running predict_particles() — preserves joint structure...")
    stat_p  = dag.predict_particles(x_prior, n_samples=3000)
    joint   = dagex.joint(stat_p)

    print()
    joint.print_summary()

    print()
    print("  Key observations:")
    r_xy = joint.correlation("x", "y")
    r_yz = joint.correlation("y", "z")
    r_zw = joint.correlation("z", "w")
    r_wo = joint.correlation("w", "out")
    print(f"    r(x, y) = {r_xy:.4f}  [x→NonLinear→y: non-monotone Mexican hat, expect moderate r]")
    print(f"    r(y, z) = {r_yz:.4f}  [y→Rectify→z: z=max(0,y), expect positive r]")
    print(f"    r(z, w) = {r_zw:.4f}  [z→Scale→w: w = 3z+1 is linear, expect r = 1.000]")
    print(f"    r(w,out)= {r_wo:.4f}  [w→Square→out: out=w², expect positive r]")
    print()
    print("  Note: predict_particles() runs node functions directly (bypasses dist_transfer),")
    print("  so the analytical Scale node contributes exact correlation just like any other node.")

    # Demonstrate assume_independent: force x ⊥ out
    joint_indep = joint.assume_independent("x", "out")
    r_before = joint.correlation("x", "out")
    r_after  = joint_indep.correlation("x", "out")
    print()
    print("  Forcing x ⊥ out (assume_independent):")
    print(f"    r(x, out) before = {r_before:.4f}")
    print(f"    r(x, out) after  = {r_after:.4f}  [≈ 0: independence enforced]")

    # Save pair plot
    import os
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    plot_path = os.path.join(os.path.dirname(__file__), "08_joint_pairs.png")
    fig, _ = joint.plot_pairs(title="Example 08 — Joint Distribution (predict_particles)")
    fig.savefig(plot_path, dpi=120, bbox_inches="tight")
    plt.close(fig)
    print(f"\n  Pair plot saved → {plot_path}")


if __name__ == "__main__":
    main()

File: examples/rs/08_predict_mc_learning.rs
// Example 08: Statistical Prediction — Monte Carlo Learning
//
// Demonstrates `Dag::predict()` when no analytical distribution transfers are
// provided.  The framework learns the output distribution by forward-sampling
// through the deterministic node functions (Monte Carlo).
//
// Pipeline
// --------
//     x ~ Normal(μ=0, σ=1)
//          │
//     [NonLinear]   y = sin(x) · exp(−x²/4)    (no closed form)
//          │
//     [Rectify]     z = max(0, y)               (no closed form)
//          │
//     [Scale]       w = 3z + 1                   (linear — attach transfer)
//          │
//     [Square]      out = w²                      (no closed form)
//
// Key concepts shown:
//   - Fully MC pass with zero transfers → all Empirical outputs
//   - Mixed-mode: analytical Scale, MC for everything else
//   - Effect of n_samples on estimate accuracy
//   - at_node early-stop for cheap intermediate inspection

mod benchmark_utils;

use std::collections::HashMap;
use std::sync::Arc;

use benchmark_utils::{print_header, print_section, Benchmark};
use dagex::{DistContext, DistTransferFn, Distribution, Graph, GraphData, PredictTarget};

// ── Node functions ────────────────────────────────────────────────────────────

fn non_linear(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let x = inputs.get("x").and_then(|d| d.as_float()).unwrap_or(0.0);
    let y = x.sin() * (-(x * x) / 4.0).exp();
    let mut out = HashMap::new();
    out.insert("y".to_string(), GraphData::float(y));
    out
}

fn rectify(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let y = inputs.get("y").and_then(|d| d.as_float()).unwrap_or(0.0);
    let mut out = HashMap::new();
    out.insert("z".to_string(), GraphData::float(y.max(0.0)));
    out
}

fn scale(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let z = inputs.get("z").and_then(|d| d.as_float()).unwrap_or(0.0);
    let mut out = HashMap::new();
    out.insert("w".to_string(), GraphData::float(z * 3.0 + 1.0));
    out
}

fn square(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let w = inputs.get("w").and_then(|d| d.as_float()).unwrap_or(0.0);
    let mut out = HashMap::new();
    out.insert("out".to_string(), GraphData::float(w * w));
    out
}

fn main() {
    print_header("Example 08: Statistical Prediction — Monte Carlo Learning");

    println!("Story:");
    println!("   A non-linear signal processing pipeline: input noise passes through");
    println!("   a Mexican-hat nonlinearity, half-wave rectification, a linear Scale,");
    println!("   and squaring.  Only Scale has a known analytical distribution transfer.");
    println!("   All other nodes are learned via MC sampling.\n");
    println!("   This demonstrates that predict() works correctly without any domain");
    println!("   knowledge — just provide the input prior and let the framework sample.\n");

    let x_prior: DistContext = HashMap::from([(
        "x".to_string(),
        Distribution::normal(0.0, 1.0),
    )]);

    // ── Build graph (pure MC — no transfers) ──────────────────────────────────
    print_section("Pure Monte Carlo (no transfers, n_samples=500)");

    let mut graph_plain = Graph::new();
    graph_plain.add(non_linear, Some("NonLinear"), Some(vec![("x", "x")]), Some(vec![("y", "y")]));
    graph_plain.add(rectify,    Some("Rectify"),   Some(vec![("y", "y")]), Some(vec![("z", "z")]));
    graph_plain.add(scale,      Some("Scale"),     Some(vec![("z", "z")]), Some(vec![("w", "w")]));
    graph_plain.add(square,     Some("Square"),    Some(vec![("w", "w")]), Some(vec![("out", "out")]));
    let dag_plain = graph_plain.build();

    let bm = Benchmark::start("MC n=500");
    let stat_plain = dag_plain.predict_at(x_prior.clone(), Some(500), None);
    let r = bm.finish();
    let s = stat_plain.summary("out").unwrap();
    println!("  out: mean={:.4}  std={:.4}  ({:.0} ms)", s.mean, s.std, r.duration_ms);

    // ── Build graph (mixed: Scale analytical + MC rest) ───────────────────────
    print_section("Mixed-mode (Scale analytical + MC elsewhere)");

    let scale_dist: DistTransferFn = Arc::new(|dists| {
        let z = dists.get("z")?;
        // w = 3z + 1  →  Normal(3μ+1, 3σ)
        Some(HashMap::from([(
            "w".to_string(),
            Distribution::normal(3.0 * z.mean() + 1.0, 3.0 * z.std()),
        )]))
    });

    let mut graph = Graph::new();
    graph.add(non_linear, Some("NonLinear"), Some(vec![("x", "x")]), Some(vec![("y", "y")]));
    graph.add(rectify,    Some("Rectify"),   Some(vec![("y", "y")]), Some(vec![("z", "z")]));
    graph.add(scale,      Some("Scale"),     Some(vec![("z", "z")]), Some(vec![("w", "w")]));
    graph.add(square,     Some("Square"),    Some(vec![("w", "w")]), Some(vec![("out", "out")]));
    graph.set_dist_transfer_for("Scale", scale_dist);
    let dag = graph.build();

    for n in [200_usize, 1000, 5000] {
        let bm = Benchmark::start("predict_at");
        let stat = dag.predict_at(x_prior.clone(), Some(n), None);
        let r = bm.finish();
        let s_out = stat.summary("out").unwrap();
        let s_w   = stat.summary("w").unwrap();
        println!(
            "  n={n:5}  out: mean={:.4}  std={:.4}  [p5={:.3}, p95={:.3}]  ({:.0} ms)",
            s_out.mean, s_out.std, s_out.p5, s_out.p95, r.duration_ms
        );
        if n == 5000 {
            println!(
                "         w (analytical Normal):  mean={:.4}  std={:.4}",
                s_w.mean, s_w.std
            );
        }
    }

    println!("\n  Note: `w` is Normal (analytical Scale transfer).");
    println!("  All other outputs are Empirical (MC samples).\n");

    // ── Inspect intermediate distributions ────────────────────────────────────
    print_section("Inspecting all intermediate distributions (n_samples=3000)");

    let stat = dag.predict_at(x_prior.clone(), Some(3000), None);

    for var in ["x", "y", "z", "w", "out"] {
        if let Some(d) = stat.get(var) {
            let s = d.summary();
            let kind = match d {
                Distribution::Normal { .. }      => "Normal",
                Distribution::Empirical { .. }   => "Empirical",
                Distribution::Deterministic(_)   => "Deterministic",
                _                                => "Other",
            };
            println!(
                "  {var:<5}  type={kind:<12}  mean={:.4}  std={:.4}  [p5={:.3}, p95={:.3}]",
                s.mean, s.std, s.p5, s.p95
            );
        }
    }

    println!();
    // Rectify clips to ≥0 so positive fraction of z should be high
    if let Some(Distribution::Empirical { samples }) = stat.get("z") {
        let positive = samples.iter().filter(|&&v| v > 0.0).count() as f64 / samples.len() as f64;
        println!("  z > 0: {:.1}%  (> 50% because rectify clips negative values to 0)", positive * 100.0);
    }

    // ── Early stop at Rectify ─────────────────────────────────────────────────
    print_section("predict_at(NodeLabel=\"Rectify\") — early stop");

    let target = PredictTarget::NodeLabel("Rectify".to_string());
    let stat_early = dag.predict_at(x_prior.clone(), Some(2000), Some(&target));

    println!("  'z' present:   {}", stat_early.contains("z"));
    println!("  'w' present:   {}", stat_early.contains("w"));
    println!("  'out' present: {}", stat_early.contains("out"));
    if let Some(s) = stat_early.summary("z") {
        println!("  z summary: mean={:.4}  std={:.4}", s.mean, s.std);
    }

    // ── n_samples convergence ─────────────────────────────────────────────────
    print_section("Effect of n_samples on output mean estimate");

    println!("  n_samples   out.mean   out.std");
    println!("  {}", "-".repeat(34));
    for n in [50_usize, 200, 1000, 5000] {
        let s = dag.predict_at(x_prior.clone(), Some(n), None).summary("out").unwrap();
        println!("  {n:8}   {:.4}   {:.4}", s.mean, s.std);
    }
    println!("\n  Estimates converge as n_samples increases.");
}
